kotest-tests/kotest-tests-core/src/jvmTest/kotlin/com/sksamuel/kotest/core/runtime/parallel/SingleInstanceTest.kt

```diff
@@ -0,0 +1,291 @@
+
+class SpecThreadSingleInstanceTest : FunSpec({
+
+   isolation = IsolationMode.SingleInstance
+   threadsForSpec = 3
+
+   val multipleThreadMultipleInvocationCounter =
+      PersistentThreadLocal<Int>()
+
+   afterSpec {
+      multipleThreadMultipleInvocationCounter.map.shouldHaveSize(3)
+      multipleThreadMultipleInvocationCounter.map.values.sum() shouldBe 3
+   }
+
+   test("test 1 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   test("test 2 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   test("test 3 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   val lock = ReentrantLock()
+
+   test("test should lock object") {
+      lock.lock()
+      delay(1000)
+      lock.unlock()
+   }
+
+   test("lock should be locked") {
+      delay(300)
+      lock.isLocked shouldBe true
+   }
+
+   test("lock should be locked too") {
+      delay(300)
+      shouldThrow<AssertionError> {
+         lock.isLocked shouldBe false
+      }
+   }
+})
+
+class SpecThreadWithNestedTestSingleInstanceTest : FunSpec({
+
+   isolation = IsolationMode.SingleInstance
+   threadsForSpec = 3
+
+   val multipleThreadCounter =
+      PersistentThreadLocal<Int>()
+
+   afterSpec {
+      assertSoftly {
+         multipleThreadCounter.map.shouldHaveSize(2)
+         multipleThreadCounter.map.values.sum() shouldBe 4
+      }
+
+   }
+
+   context("First single thread context") {
+      test("test 1 should create new key in map for context or add value 1") {
+         val counter = multipleThreadCounter.getOrSet { 0 }
+         multipleThreadCounter.set(counter + 1)
+      }
+
+      test("test 2 should create new key in map for context or add value 1") {
+         val counter = multipleThreadCounter.getOrSet { 0 }
+         multipleThreadCounter.set(counter + 1)
+      }
+   }
+
+   context("Second single thread context") {
+      test("test 1 should create new key in map for context or add value 1") {
+         val counter = multipleThreadCounter.getOrSet { 0 }
+         multipleThreadCounter.set(counter + 1)
+      }
+
+      test("test 2 should create new key in map for context or add value 1") {
+         val counter = multipleThreadCounter.getOrSet { 0 }
+         multipleThreadCounter.set(counter + 1)
+      }
+   }
+
+   val lock = ReentrantLock()
+
+   context("Third single thread context") {
+
+      test("test should lock object") {
+         lock.lock()
+         delay(1000)
+         lock.unlock()
+      }
+
+      test("lock should be unlocked") {
+         delay(300)
+         lock.isLocked shouldBe false
```

Comment:
Please explain this test to me.

Reply:
This test assert that all nested tests (all tests from the same context) run sequential by one thread.
If compare this tests with similar tests from `InstancePerTestTest.kt` or `InstancePerLeafTest.kt` you will see difference.